rs/bitcoin/kyt/src/fetch.rs

```diff
@@ -0,0 +1,270 @@
+}
+
+pub trait TransactionLike {
+    type Input;
+    type AddressError;
+
+    fn iter_inputs(&self) -> impl Iterator<Item = &Self::Input>;
+    fn output_address(&self, vout: u32) -> Result<Address, Self::AddressError>;
+}
+
+/// Trait that abstracts over state operations.
+pub trait FetchState<T> {
+    type FetchGuard;
+    fn new_fetch_guard(&self, txid: Txid) -> Result<Self::FetchGuard, FetchGuardError>;
+    fn get_fetch_status(&self, txid: Txid) -> Option<FetchTxStatus<T>>;
+    fn set_fetch_status(&self, txid: Txid, status: FetchTxStatus<T>);
+    fn set_fetched_address(&self, txid: Txid, index: usize, address: Address);
+}
+
+/// Trait that abstracts over system functions like fetching transaction, calcuating cycles, etc.
+pub trait FetchEnv<I: HasOutPoint, T: Clone + TransactionLike<Input = I>>
+where
+    <T as TransactionLike>::AddressError: Debug,
+    T: Debug,
+{
+    async fn get_tx(&self, txid: Txid, buffer_size: u32) -> Result<T, GetTxError>;
+    fn cycles_accept(&self, cycles: u128) -> u128;
+    fn cycles_available(&self) -> u128;
+
+    /// Try to fetch a transaction given its txid:
+    /// - If it is already available, return `Fetched`.
+    /// - If it is already pending, return `Pending`.
+    /// - If it is pending retry or not found, return a future that calls `fetch_tx`.
+    /// - Or return other conditions like `HighLoad` or `Error`.
+    fn try_fetch_tx<State: FetchState<T>>(
+        &self,
+        state: &State,
+        txid: Txid,
+    ) -> TryFetchResult<T, impl futures::Future<Output = Result<FetchResult<T>, ()>>> {
+        let buffer_size = match state.get_fetch_status(txid) {
+            None => INITIAL_BUFFER_SIZE,
+            Some(FetchTxStatus::PendingRetry { buffer_size, .. }) => buffer_size,
+            Some(FetchTxStatus::PendingOutcall { .. }) => return TryFetchResult::Pending,
+            Some(FetchTxStatus::Error(msg)) => return TryFetchResult::Error(msg),
+            Some(FetchTxStatus::Fetched(fetched)) => return TryFetchResult::Fetched(fetched),
+        };
+        let guard = match state.new_fetch_guard(txid) {
+            Ok(guard) => guard,
+            Err(_) => return TryFetchResult::HighLoad,
+        };
+        let cycle_cost = get_tx_cycle_cost(buffer_size);
+        if self.cycles_accept(cycle_cost) < cycle_cost {
+            TryFetchResult::NotEnoughCycles
+        } else {
+            TryFetchResult::ToFetch(self.fetch_tx(state, guard, txid, buffer_size))
+        }
+    }
+
+    /// Fetch a transaction using http outcall by its txid and set its status to:
+    /// - `Fetched`, if it is available.
+    /// - `PendingRetry`, if the allocated buffer for outcall wasn't enough.
+    /// - `Error`, if an irrecoverable error happened during the outcall of `get_tx`.
+    ///
+    /// Return the correponding `FetchResult`.
+    ///
+    /// Note that this function does not return any error, but due to requirements
+    /// of `try_join_all` it must return a `Result` type.
```

Comment:
I don't understand this part. If there is no error, can't you extract the result and return `FetchResult<T>`?

Reply:
`try_join_all` insists that all futures it joins together must give a `Result` type. So this is doing just that, but filling the `Err` part with `()`, which may look a bit strange. So I was trying to explain this because the futures by design really should not return `Err` at all.

Reply:
Good suggestion! TIL!